File: 29_08_2024_Most_Stones_Removed_with_Same_Row_or_Column.java
import java.util.List;
import java.util.ArrayList;

class Solution {
    private int connectedComponentCount = 0;

    private int findRepresentative(int element, int[] setRepresentatives) {
        if (setRepresentatives[element] == 0) {
            setRepresentatives[element] = element;
            connectedComponentCount++;
        }
        return setRepresentatives[element] == element ? element : 
               (setRepresentatives[element] = findRepresentative(setRepresentatives[element], setRepresentatives));
    }

    private void mergeComponents(int elementA, int elementB, int[] setRepresentatives) {
        int repA = findRepresentative(elementA, setRepresentatives);
        int repB = findRepresentative(elementB, setRepresentatives);
        if (repA != repB) {
            setRepresentatives[repB] = repA;
            connectedComponentCount--;
        }
    }

    public int removeStones(List<List<Integer>> stonePositions) {
        int[] setRepresentatives = new int[20003];
        for (List<Integer> stonePosition : stonePositions) {
            mergeComponents(stonePosition.get(0) + 1, stonePosition.get(1) + 10002, setRepresentatives);
        }
        return stonePositions.size() - connectedComponentCount;
    }
}

File: 29_08_2024_Most_Stones_Removed_with_Same_Row_or_Column.py
class Solution:
    def __init__(self):
        self.connected_component_count = 0

    def find_representative(self, element, set_representatives):
        if set_representatives[element] == 0:
            set_representatives[element] = element
            self.connected_component_count += 1
        return set_representatives[element] if set_representatives[element] == element else \
               (set_representatives[element] := self.find_representative(set_representatives[element], set_representatives))

    def merge_components(self, element_a, element_b, set_representatives):
        rep_a = self.find_representative(element_a, set_representatives)
        rep_b = self.find_representative(element_b, set_representatives)
        if rep_a != rep_b:
            set_representatives[rep_b] = rep_a
            self.connected_component_count -= 1

    def remove_stones(self, stone_positions):
        set_representatives = [0] * 20003
        for stone_position in stone_positions:
            self.merge_components(stone_position[0] + 1, stone_position[1] + 10002, set_representatives)
        return len(stone_positions) - self.connected_component_count

